tangos/input_handlers/pynbody.py

```diff
@@ -439,21 +440,86 @@ def _is_able_to_load(self, filepath):
             return False
 
 
-class GadgetAHFInputHandler(PynbodyInputHandler):
-    patterns = ["snapshot_???"]
-    auxiliary_file_patterns = ["*.AHF_particlesSTARDUST"]
 
-    def _is_able_to_load(self, filepath):
-        try:
-            f = pynbody.load(filepath)
-            h = pynbody.halo.AHFCatalogue(f)
-            return True
-        except (IOError, RuntimeError):
-            return False
+class AHFInputHandler(PynbodyInputHandler):
+    pynbody_halo_class_name = "AHFCatalogue"
+
+    _included_additional_properties = (
+        "parent", "child", "shrink_center", "bulk_velocity",
+    )
+
+    _excluded_precalculated_properties = (
+        "boxsize", "time", "hostHalo", "Xc", "Yc", "Zc",
+        "VXc", "VYc", "VZc",
+    )
+
+    def available_object_property_names_for_timestep(self, ts_extension, object_typetag):
+        h = self._construct_halo_cat(ts_extension, object_typetag)
+
+        return (
+            [
+                key for key in h[1].properties.keys() 
+                if key not in self._excluded_precalculated_properties
+            ] + list(self._included_additional_properties)
+        )
+        
 
+    def _get_map_child_subhalos(self, ts_extension):
+        h = self._construct_halo_cat(ts_extension, 'halo')
+        halo_children = defaultdict(list)
+        for halo in h:
+            iparent = halo.properties["hostHalo"]
+            if iparent == 0:
+                continue
+
+            halo_children[iparent].append(halo.properties["halo_id"])
+        return halo_children
+
+    def iterate_object_properties_for_timestep(self, ts_extension, object_typetag, property_names):
+        h = self._construct_halo_cat(ts_extension, object_typetag)
+
+        h.physical_units()
+
+        if "child" in property_names:
+            map_child_parent = self._get_map_child_subhalos(ts_extension)
+
+        for halo in h:
+            # Tangos expects data to have a finder offset, and a finder id following the stat file logic
```

Comment:
AHF 'offsets' and 'ids' can be different -- this is actually why this is here. I think it only becomes a problem when AHF is run with MPI. @mtremmel almost certainly has a comment on this...

Reply:
I see. Is this the difference between `ID` and `halo_id` in AHF properties? These two quantities are present in the catalog and I'm not entirely sure what they map onto

Reply:
Hi sorry for the late response on this. Yes these two may be different. finder_offset encodes the position of the halo within the catalog, which is what pynbody uses directly. For example, load_copy(10) looks for the 10th halo in the catalog and loads it. It used to be always the case that the Nth halo ID in AHF files was always N - 1 (i.e. the 10th halo would be ID number 9). However, many halo finders, AHF included, have at least the option to produce completely random unique ID numbers for their halos that are completely disconnected from their position within the catalog. So, now we separate these out. Finder_ID reads the ID number directly from the catalog while finder_offset encodes the order within the catalog. Pynbody uses the latter, but we wanted to remain agnostic that other codes may use either identification number.

Reply:
If I interpret correctly what you are doing here, you are explicitly assuming that both finder_offset and finder_id are the same and equal to the ID number in the catalog. This is probably true if you didn't run with MPI as @apontzen mentioned, but it doesn't have to be true (this is actually a parameter you can set in AHF when you run it with or without MPI).

Reply:
hmmm I'm a bit confused here. What do you mean by outputted by AHF? I'm looking at an AHF output I have for the halo stat file (the one that lists all the halo properties) and I only see ID (the first value). There is also a "halo_id" column in yours? I looked at the AHF website and I didn't see any change in the documentation, though it is possible there's been an update to the format I don't know about

Reply:
to be clear, I'm referring to the file ending in ".AHF_halos"

Reply:
I assume AHF halo stat files are usually generated via pynbody in your workflow? From these lines:
https://github.com/pynbody/pynbody/blob/ee8fb2f637006e16f189ce2c6c92259daf88d4d9/pynbody/halo/ahf.py#L470-L487

it looks like that `halo_id` maps onto the what would be written under the `Grp` header of the stat file

Reply:
ah sorry right, you are using pynbody to read in the .AHF_halos file essentially and it does this for you. Here 'ID' is taken directly from the AHF output (so this would  indeed be finder_id) and 'halo_id' is how pynbody tracks the halo (1 = first, 2 = second, etc), so that should be your finder_offset in tangos. You are right!

Reply:
Oh, I see. No I never use the postprocessed halo catalog like that, only the original AHF catalog. TANGOS interacts with this as well with the HaloStatFile class. Sorry I was confused. However, I'm also not sure how this postprocessed stat file would interact with an AHF run that uses the random ID numbers

Reply:
Ok @cphyc @apontzen @mtremmel, I added a more detailed comment of the situation, which I am still not fully cleared up on (turns out the code didn't need to change as far as I understand). 

I don't think this is particularly ideal, but the `PropertyImporter` needs fixing in more general way, and it has been modified in a way that is likely to break other handlers. I think this is beyond the scope of this PR, and that the best way is to open an issue from this discussion and fix in a different PR. Thoughts?

Reply:
I'm not actually sure what the problem is at the moment so I can't really comment. I just noted the inconsistency. But in particular I'm not sure what is wrong in `PropertyImporter` currently?  Are you saying it's currently broken in the main branch in some way?

Reply:
It's not broken per se, mostly inconsistent in an undocumented way? The expected behaviour of the enumeration has been changed by PR #88, from receiving one number to receiving two numbers before the actual properties to be committed to the database (commit ea7c1753).

This might be ok for workflows built upon stat files, but for handlers enumerating properties through pynbody, I am unclear what the second property is meant to be. 

Furthermore, I think earlier pynbody handlers were not updated to reflect the change in behaviour, by inserting a dummy second number before `yield`. For example, the `GadgetSubfindHandler` look like it would be missing one property following the enumeration here:
https://github.com/pynbody/tangos/blob/b8283572e1d2181df0980a6dd665cb921da00872/tangos/input_handlers/pynbody.py#L392-L424

Reply:
Yeah the current code was basing this on AHF output, not on pynbody's post processed halo catalog outputs which is where it seems like the problem is arising?

Reply:
Ok thanks. I have updated the comment to be explicit about the behaviour, and the documentation of the base class to reflect the new expectations. 

I will open a separate issue to flag that (I think) old pynbody handlers might have been broken by the change and need a one-line update.